### PR TITLE
fix(routing): snapping query points on node instead of edges

### PR DESCRIPTION
**Pull request by kschrab**
## Type of this PR 

- [x] Bug fix
- [x] Enhancement

## Description

* When passing a `RoutePosition` as source or target for a routing query, the implementation first tries to match the given position onto the road, by either snapping it to a node (`Position.TOWER`) or connection (`Position.EDGE`). In some rare cases, the queried position was mapped onto a connection, although the position was very close to a node (few millimeters). To fix those cases, we now switch the snapping mode of the `QueryResult` to `Position.TOWER`, if the queried position is very close to the closest point of the matched edge. The distance to control this behavior can be adjusted with the `TARGET_NODE_QUERY_DISTANCE` field.
* Added a new method to `RoutingParameters` to define how much costs are applied for a turn restrictions. Setting this value allows to ignore turn-restrictions to some extend (e.g. for demo cases).

## Issue(s) related to this PR

 * Internal issue 320 
 
## Affected parts of the online documentation

No

## Definition of Done

- [x] You have read CONTRIBUTING.md carefully.
- [x] You have signed the [Contributor License Agreement](http://www.eclipse.org/legal/CLA.php).
- [x] All commits are signed-off (`-s`) with your mail address of your Eclipse Account.
- [x] `origin/main` has been merged into your Fork.
- [x] New functionality is covered by unit tests or integration tests. Code coverage must not decrease.
- [x] There are no new SpotBugs warnings. 
- [x] Coding guidelines have been observed (see CONTRIBUTING.md).
- [x] All tests pass.

### CODE REVIEW

File: lib/mosaic-routing/src/main/java/org/eclipse/mosaic/lib/routing/graphhopper/GraphHopperRouting.java [hunk 1]
```diff
@@ -127,69 +135,100 @@ public GraphHopperRouting loadGraphFromDatabase(Database db) {
...
+        } else {
+            return fixQueryResultIfSnappedPointIsCloseToTowerNode(qrTo, target);
+        }
+    }
+
+    private QueryResult createQueryForSource(FlagEncoder flagEncoder, RoutingPosition source) {
+        final EdgeFilter fromEdgeFilter = createEdgeFilterForRoutingPosition(source, flagEncoder);
+        QueryResult qrFrom = ghApi.getLocationIndex().findClosest(source.getPosition().getLatitude(), source.getPosition().getLongitude(), fromEdgeFilter);
+        if (source.getConnectionId() != null) {
+            qrFrom = fixQueryResultIfSnappedPointIsTowerNode(qrFrom, source, fromEdgeFilter);
```
Comment (schwepmo) [inline lib/mosaic-routing/src/main/java/org/eclipse/mosaic/lib/routing/graphhopper/GraphHopperRouting.java:197]: As far as I see this is the only difference from `creqteQueryForTarget`, should we unify the two methods?
Also is there a reason why we wouldn't snap the target position to a Tower Node?

File: lib/mosaic-routing/src/main/java/org/eclipse/mosaic/lib/routing/graphhopper/GraphHopperRouting.java [hunk 2]
```diff
@@ -226,7 +265,8 @@ private EdgeFilter createEdgeFilterForRoutingPosition(final RoutingPosition posi
      */
     private Node determinePrefixNode(final RoutingPosition source) {
         if (source.getConnectionId() != null) {
-            return db.getConnection(source.getConnectionId()).getFrom();
+            List<Node> nodes = db.getConnection(source.getConnectionId()).getNodes();
+            return nodes.get(Math.max(0, nodes.size() - 2));
```
Comment (schwepmo) [inline lib/mosaic-routing/src/main/java/org/eclipse/mosaic/lib/routing/graphhopper/GraphHopperRouting.java:269]: What's the reasoning to use this node index?

**Review by schwepmo — COMMENTED** [review]
I left some comments. No major issues.

File: lib/mosaic-routing/src/main/java/org/eclipse/mosaic/lib/routing/graphhopper/GraphHopperRouting.java (hunk 1 above)
Comment (kschrab) [reply lib/mosaic-routing/src/main/java/org/eclipse/mosaic/lib/routing/graphhopper/GraphHopperRouting.java:197]: This is a special treatment for the source when we are doing route request from applications, as we **always** require to match the edge the vehicle is currently on. Otherwise the calculated routes can not be used with SUMO. For the target this is not explicitly required. This is also related to your other comment: we want to find an edge, and it is much more likely to match an edge if we search with a position which is located ON the connection, and not at the beginning of it (as it was before). This was a leftover when we unified edges and connections.

Tl;dr: we could also do this for the target, but I cannot estimate the implications of this right now.

File: lib/mosaic-routing/src/main/java/org/eclipse/mosaic/lib/routing/graphhopper/GraphHopperRouting.java (hunk 2 above)
Comment (kschrab) [reply lib/mosaic-routing/src/main/java/org/eclipse/mosaic/lib/routing/graphhopper/GraphHopperRouting.java:269]: I'll add a comment. Actually we have more than 2 nodes per connection in the usual case.
